### DISCUSSION TITLE
Uploading Files From NextJs to Supabase Storage 403 error

### DISCUSSION THREAD

**Discussion by LinThitHtwe** [supabase/supabase#26291] [Questions] [answered]
Hello, I want to upload my files to supabase  but I've been getting 403 error. I am having a hard time understanding Access control ,RLS and other stuffs in documentation. Can anyone help me out ? TwT.

Can I just allow anyone to insert, update, delete without authentication, just for the testing purpose ?
Or if I add authentication and authorization , how can I do it ?

``` typescript
const { data, error } = await supabase.storage
    .from("Test")
    .upload(file.name, file);
  if (error) {
  //....
  }
  if (data) {
  //....
  }

```

This is my error 
```javascript
{
  statusCode: '403',
  error: 'Unauthorized',
  message: 'new row violates row-level security policy'
}
```

**Comment by silentworks** [comment, marked as answer]
You can set it to allow write to all storage buckets and objects by adding a policy to each and setting its value to true. You can also do this via the SQL editor using the sql below. Do note we do not advice on doing this on a production app or any public facing project, you should always configure your RLS correctly and not leave them open like this.

```sql
CREATE POLICY "Enable read access for all users" ON "storage"."objects"
AS PERMISSIVE FOR SELECT
TO public
USING (true);

CREATE POLICY "Enable insert for all users" ON "storage"."objects"
AS PERMISSIVE FOR INSERT
TO public
WITH CHECK (true);

CREATE POLICY "Enable update for all users" ON "storage"."objects"
AS PERMISSIVE FOR UPDATE
TO public
USING (true)
WITH CHECK (true);

CREATE POLICY "Enable delete for all users" ON "storage"."objects"
AS PERMISSIVE FOR DELETE
TO public
USING (true);

CREATE POLICY "Enable read access for all users" ON "storage"."buckets"
AS PERMISSIVE FOR SELECT
TO public
USING (true);

CREATE POLICY "Enable insert for all users " ON "storage"."buckets"
AS PERMISSIVE FOR INSERT
TO public
WITH CHECK (true);

CREATE POLICY "Enable update for all users" ON "storage"."buckets"
AS PERMISSIVE FOR UPDATE
TO public
USING (true)
WITH CHECK (true);

CREATE POLICY "Enable delete for all users" ON "storage"."buckets"
AS PERMISSIVE FOR DELETE
TO public
USING (true);
```

**Reply by LinThitHtwe** [reply]
Thanks a lot. It did work now.
May I know how can I restrict as well ? 
Like I only want me and just me to insert, update and delete the data from storage bucket?

Is there other sources would you like to suggest me to learn more about storages and buckets in supabase? 
I'm having a hard time understanding these by reading documentations. 

ChatGPT suggested me with this but I have no idea what's going on
```sql
CREATE POLICY "Enable read access for all users" ON "storage"."objects"
AS PERMISSIVE FOR SELECT
TO public
USING (true);

CREATE POLICY "Enable insert for specific user" ON "storage"."objects"
AS PERMISSIVE FOR INSERT
TO my_name
WITH CHECK (true);

CREATE POLICY "Enable update for specific user" ON "storage"."objects"
AS PERMISSIVE FOR UPDATE
TO my_name
USING (true)
WITH CHECK (true);

CREATE POLICY "Enable delete for specific user" ON "storage"."objects"
AS PERMISSIVE FOR DELETE
TO my_name
USING (true);

-- Policies for the "storage"."buckets" table
CREATE POLICY "Enable read access for all users" ON "storage"."buckets"
AS PERMISSIVE FOR SELECT
TO public
USING (true);

CREATE POLICY "Enable insert for specific user" ON "storage"."buckets"
AS PERMISSIVE FOR INSERT
TO my_name
WITH CHECK (true);

CREATE POLICY "Enable update for specific user" ON "storage"."buckets"
AS PERMISSIVE FOR UPDATE
TO my_name
USING (true)
WITH CHECK (true);

CREATE POLICY "Enable delete for specific user" ON "storage"."buckets"
AS PERMISSIVE FOR DELETE
TO my_name
USING (true);


```

**Comment by LinThitHtwe** [comment]
> You can set it to allow write to all storage buckets and objects by adding a policy to each and setting its value to true. You can also do this via the SQL editor using the sql below. Do note we do not advice on doing this on a production app or any public facing project, you should always configure your RLS correctly and not leave them open like this.
> 
> ```sql
> CREATE POLICY "Enable read access for all users" ON "storage"."objects"
> AS PERMISSIVE FOR SELECT
> TO public
> USING (true);
> 
> CREATE POLICY "Enable insert for all users" ON "storage"."objects"
> AS PERMISSIVE FOR INSERT
> TO public
> WITH CHECK (true);
> 
> CREATE POLICY "Enable update for all users" ON "storage"."objects"
> AS PERMISSIVE FOR UPDATE
> TO public
> USING (true)
> WITH CHECK (true);
> 
> CREATE POLICY "Enable delete for all users" ON "storage"."objects"
> AS PERMISSIVE FOR DELETE
> TO public
> USING (true);
> 
> CREATE POLICY "Enable read access for all users" ON "storage"."buckets"
> AS PERMISSIVE FOR SELECT
> TO public
> USING (true);
> 
> CREATE POLICY "Enable insert for all users " ON "storage"."buckets"
> AS PERMISSIVE FOR INSERT
> TO public
> WITH CHECK (true);
> 
> CREATE POLICY "Enable update for all users" ON "storage"."buckets"
> AS PERMISSIVE FOR UPDATE
> TO public
> USING (true)
> WITH CHECK (true);
> 
> CREATE POLICY "Enable delete for all users" ON "storage"."buckets"
> AS PERMISSIVE FOR DELETE
> TO public
> USING (true);
> ```

I got it with auth. Thanks.
